cloudinit/net/dhcp.py

```diff
@@ -503,62 +540,254 @@ def get_latest_lease(lease_dir, lease_file_regex):
+    def dhcp_discovery(
+        self,
+        interface: str,
+        dhcp_log_func: Optional[Callable] = None,
+        distro=None,
+    ) -> Dict[str, Any]:
+        """Run dhcpcd on the interface without scripts/filesystem artifacts.
+
+        @param interface: Name of the network interface on which to send a
+            dhcp request
+        @param dhcp_log_func: A callable accepting the client output and
+            error streams.
+        @param distro: a distro object for network interface manipulation
+        @return: dict of lease options representing the most recent dhcp lease
+            parsed from the dhclient.lease file
+        """
+        LOG.debug("Performing a dhcp discovery on %s", interface)
+        sleep_time = 0.01
+        sleep_cycles = int(self.timeout / sleep_time)
+
+        # ISC dhclient needs the interface up to send initial discovery packets
+        # Generally dhclient relies on dhclient-script PREINIT action to bring
+        # the link up before attempting discovery. Since we are using
+        # -sf /bin/true, we need to do that "link up" ourselves first.
+        distro.net_ops.link_up(interface)
+
+        # TODO: disabling hooks means we need to get all of the files in
+        # /lib/dhcpcd/dhcpcd-hooks/ and pass each of those with the --nohook
+        # argument to dhcpcd
+        try:
+            command = [
+                self.dhcp_client_path,  # pyright: ignore
+                "--ipv4only",  # only attempt configuring ipv4
+                "--waitip",  # wait for ipv4 to be configured
+                "--persistent",  # don't deconfigure when dhcpcd exits
+                "--noarp",  # don't be slow
+                "--script=/bin/true",  # disable hooks
+                interface,
+            ]
+            out, err = subp.subp(
+                command,
+                timeout=self.timeout,
+            )
+            if dhcp_log_func is not None:
+                dhcp_log_func(out, err)
+            lease = self.get_newest_lease(interface)
+            # Attempt cleanup and leave breadcrumbs if it fails, but return
+            # the lease regardless of failure to clean up dhcpcd.
+            if lease:
+                # Note: the pid file location depends on the arguments passed
+                # it can be discovered with the -P flag
+                pid_file = subp.subp([*command, "-P"]).stdout
+                pid_content = None
+                debug_msg = ""
+                for _ in range(sleep_cycles):
+                    try:
+                        pid_content = util.load_file(pid_file).strip()
+                        pid = int(pid_content)
+                        LOG.debug("killing dhcpcd with pid=%s", pid)
+                        os.kill(pid, signal.SIGKILL)
+                    except FileNotFoundError:
+                        debug_msg = (
+                            f"No PID file found at {pid_file}, "
+                            "dhcpcd is still running"
```

Comment:
We are running into this condition on ec2 where we loop 1000 times over 10 seconds before finally giving up and leaving dhcpcd running.

Reply:
Hrm, I saw this previously, but that was before I had the `sleep()` call there - I haven't seen this yet since adding it in. Back when I was debugging that I noticed that child PID gets printed to standard out, and had written some code that parsed and killed the child process using that. I just pushed that code, as well as changes to address your previous comments.

Reply:
Since parsing the lease file from stdin is [currently unreliable](https://github.com/NetworkConfiguration/dhcpcd/issues/285), and `dhcpcd --dumplease -4 <ifname>`  expects to be able to talk to a running daemon, and dumping all leases also expects to talk to a running daemon (and [is unreliable](https://github.com/NetworkConfiguration/dhcpcd/issues/287)), I think that we're stuck trying to use a daemon and clean it up for now. Once https://github.com/NetworkConfiguration/dhcpcd/issues/285 is fixed we can change that.

In the meantime, the only alternative that I've thought of involves the `ptrace` syscall. I want to leave the direct use of the syscall out of scope, since that would require architecture-dependent (and probably kernel-dependent) code via ctypes, but perhaps calling `dhcpcd` under `strace --follow-forks --trace=process <dhcpcd call>` or `strace --follow-forks -e trace="clone,fork,vfork"` would be an acceptable stopgap until https://github.com/NetworkConfiguration/dhcpcd/issues/287 is fixed. Unfortunately `strace` isn't going to be available everywhere, so this would have to be optional depending on its availability. It might raise some eyebrows, but it would probably "work".

Reply:
Example of using strace to get PID.
```
sudo strace --daemonize=grandchild --follow-forks --trace=clone,fork,vfork,clone --signal='!all' dhcpcd --ipv4only --waitip=4 --persistent --noarp enp24s0
strace: Process 133407 attached
dhcpcd-9.4.1 starting
clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133412 attached
, child_tidptr=0x7f73cef02a10) = 133412
[pid 133412] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133413 attached
, child_tidptr=0x7f73cef02a10) = 133413
[pid 133412] +++ exited with 0 +++
[pid 133413] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD, child_tidptr=0x7f73cef02a10) = 133414
strace: Process 133414 attached
[pid 133413] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133415 attached
, child_tidptr=0x7f73cef02a10) = 133415
DUID 00:04:00:00:00:00:00:00:00:00:00:00:30:9c:23:e6:be:f0
strace: Process 133416 attached
[pid 133416] +++ exited with 0 +++
strace: Process 133417 attached
[pid 133417] +++ exited with 0 +++
enp24s0: IAID 23:e6:be:f0
enp24s0: rebinding lease of 192.168.0.104
[pid 133414] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133418 attached
, child_tidptr=0x7f73cef02a10) = 133418
[pid 133414] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133419 attached
, child_tidptr=0x7f73cef02a10) = 133419
enp24s0: leased 192.168.0.104 for 86400 seconds
enp24s0: adding route to 192.168.0.0/24
enp24s0: adding default route via 192.168.0.1
strace: Process 133420 attached
[pid 133420] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133421 attached
, child_tidptr=0x7f3d38fa2a10) = 133421
[pid 133421] +++ exited with 0 +++
[pid 133420] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133422 attached
, child_tidptr=0x7f3d38fa2a10) = 133422
[pid 133422] +++ exited with 0 +++
[pid 133420] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133423 attached
, child_tidptr=0x7f3d38fa2a10) = 133423
[pid 133420] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133424 attached
, child_tidptr=0x7f3d38fa2a10) = 133424
[pid 133423] +++ exited with 0 +++
Dropped protocol specifier '.dhcp' from 'enp24s0.dhcp'. Using 'enp24s0' (ifindex=2).
[pid 133424] +++ exited with 0 +++
[pid 133420] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLDstrace: Process 133427 attached
, child_tidptr=0x7f3d38fa2a10) = 133427
[pid 133427] +++ exited with 0 +++
[pid 133420] +++ exited with 0 +++
forked to background, child pid 133413
[pid 133414] clone(child_stack=NULL, flags=CLONE_CHILD_CLEARTID|CLONE_CHILD_SETTID|SIGCHLD <unfinished ...>
[pid 133407] +++ exited with 0 +++
```

Reply:
This raises a good point however that we should be killing all of the processes, not just one of them. Killing by process group would accomplish that. See the latest commit pushed @blackboxsw.

Reply:
Some points on your ocmments above:
1. > I think that we're stuck trying to use a daemon and clean it up for now."

 Agreed, because of the former bugs, we should spawn and cleanup dhcpcd proceses. Please link in a comment relative to the dhcpcd command the upstream bugs prompting this behavior so we can track the feature/fix in the future and re-evaluate our solution

2. Agreed to stay away from strrace to determine PIDs as this is highly environment dependent and likely fragile. 

3. > In the meantime, the only alternative that I've thought of involves the ptrace syscall. I want to leave the direct use of the syscall out of scope, since that would require architecture-dependent

Why do we need ptrace at all. We can/do obtain the ppid of the dhcpcd process from /run/dhcpcd/*.pid properly which allows us to directly kill that process and any children with a `os.kill(<pid_from_dhcpcd_pid_file>)`


Given the `stdout.strip()` fix to the load_file(pid_file). It looks like the code you have performs the right operation and could kill the dhcpcd parent pid cleanly without all the /proc/<PID>/stat lookup functions.